### DISCUSSION TITLE
[Duplicate] Hide ReferenceEquals and static Equals

### DISCUSSION THREAD

**Discussion by YegorStepanov** [dotnet/roslyn#59420] [General] [answered]
Originally posted in https://github.com/dotnet/roslyn/discussions/59333 with the answer:
> Due to risk of breaking valid scenarios and lack of requests, we will almost certainly not be implementing this request.

Since this is quite fundamental, I think I can post here.

# Introduction

It's annoying to see __ReferenceEquals__ and __Equals__ anywhere.

![pikcha](https://user-images.githubusercontent.com/19392641/153239175-2c435e76-e900-4af1-a610-b3c4d48c72b4.png)

## Static classes
Both methods are meaningless:
```C#
Math.Equals(...);
Math.ReferenceEquals(...);
```

`Assert`, Serilog `Log` and small user's helper classes suffer from this at code completion.

## Struct
The static __Equals__ is _wrong_ for any struct.
```C#
int.Equals(3, 3);
```

This bug easily appears when changing a class to a struct in a user type.

## ReferenceEquals
__ReferenceEquals__ is used via object or via direct call. I've never seen the code like:
```C#
string.ReferenceEquals(...);
```

Also this is a low-level method, but you see it at every code completion for a type.

# Proposal
Hide static __Equals__ for structs and static classes.    
Hide __ReferenceEquals__ for all types, except __object__.

This should be done by hiding them from IntelliSense. The new analyzer will complain about the use of these hidden methods.
Not sure if these methods should be hidden from Reflection.

# Additional options
1) Hide direct __Equals__ call for structs? It will help to better detect the "changing a class to a struct" bug.

2) Hide direct __ReferenceEquals__ call? 
* You will have to write `object.ReferenceEquals(...)` always. 
* It will unifies code, for example dotnet/runtime uses both.
* This will completely hide the infrastructure method in __object__.
* cons: there will be a lot of warnings as direct __ReferenceEquals__ is used often.

**Comment by HaloFour** [comment]
This sounds like a tooling suggestion and not a language suggestion.  Maybe try opening on the [Roslyn repo](https://github.com/dotnet/roslyn/issues)?

**Reply by YegorStepanov** [reply]
I've opened it there, but
>> lack of requests

This repo is more popular.

For me it's not much, but it changes the feel of the language.

**Reply by CyrusNajmabadi** [reply]
> This repo is more popular.

That's not relevant.  CSharplang is *only* for the C# language.  This is out of scope.  I'm moving this to dotnet/roslyn.

**Comment by CyrusNajmabadi** [comment]
Moving to roslyn due to:

> This should be done by hiding them from IntelliSense.

> The new analyzer will complain about the use of these hidden methods.

These are not language concerns. These are both tooling concerns.

**Reply by YegorStepanov** [reply]
Sorry for the misunderstanding.

The part of proposal for static classes said to unbind static classes from `System.Object` in soft form. Is it can theoretically be done silently in roslyn repo?

**Reply by CyrusNajmabadi** [reply]
> The part of proposal for static classes said to unbind static classes from System.Object in soft form. Is it can theoretically be done silently in roslyn repo?

I'm not sure what you're asking for.  However, you could certainly write your own analyzer to find this and force you to stop.  You could also suggest such an analyzer at github.com/dotnet/roslyn-analyzers.

**Comment by sharwell** [comment, marked as answer]
Duplicate of #59333
